cli/cliui/table_test.go

```diff
@@ -0,0 +1,346 @@
+package cliui_test
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/coder/coder/cli/cliui"
+)
+
+type tableTest1 struct {
+	Name        string      `table:"name"`
+	NotIncluded string      // no table tag
+	Age         int         `table:"age"`
+	Roles       []string    `table:"roles"`
+	Sub1        tableTest2  `table:"sub 1,recursive"`
+	Sub2        *tableTest2 `table:"sub 2,recursive"`
+	Sub3        tableTest3  `table:"sub 3,recursive"`
+
+	// Types with special formatting.
+	Time    time.Time  `table:"time"`
+	TimePtr *time.Time `table:"time ptr"`
+}
+
+type tableTest2 struct {
+	Name        string `table:"name"`
+	Age         int    `table:"age"`
+	NotIncluded string // no table tag
+}
+
+type tableTest3 struct {
+	NotIncluded string     // no table tag
+	Sub         tableTest2 `table:"inner,recursive"`
+}
+
+func Test_DisplayTable(t *testing.T) {
+	t.Parallel()
+
+	// This test tests skipping fields without table tags, recursion, pointer
+	// dereferencing, and nil pointer skipping.
+	t.Run("OK", func(t *testing.T) {
+		t.Parallel()
+
+		someTime := time.Date(2022, 8, 2, 15, 49, 10, 0, time.Local)
+		in := []tableTest1{
+			{
+				Name:  "foo",
+				Age:   10,
+				Roles: []string{"a", "b", "c"},
+				Sub1: tableTest2{
+					Name: "foo1",
+					Age:  11,
+				},
+				Sub2: &tableTest2{
+					Name: "foo2",
+					Age:  12,
+				},
+				Sub3: tableTest3{
+					Sub: tableTest2{
+						Name: "foo3",
+						Age:  13,
+					},
+				},
+				Time:    someTime,
+				TimePtr: &someTime,
+			},
+			{
+				Name:  "bar",
+				Age:   20,
+				Roles: []string{"a"},
+				Sub1: tableTest2{
+					Name: "bar1",
+					Age:  21,
+				},
+				Sub2: nil,
+				Sub3: tableTest3{
+					Sub: tableTest2{
+						Name: "bar3",
+						Age:  23,
+					},
+				},
+				Time:    someTime,
+				TimePtr: nil,
+			},
+			{
+				Name:  "baz",
+				Age:   30,
+				Roles: nil,
+				Sub1: tableTest2{
+					Name: "baz1",
+					Age:  31,
+				},
+				Sub2: nil,
+				Sub3: tableTest3{
+					Sub: tableTest2{
+						Name: "baz3",
+						Age:  33,
+					},
+				},
+				Time:    someTime,
+				TimePtr: nil,
+			},
+		}
+
+		expected := `
+			NAME  AGE  ROLES    SUB 1 NAME  SUB 1 AGE  SUB 2 NAME  SUB 2 AGE  SUB 3 INNER NAME  SUB 3 INNER AGE  TIME             TIME PTR
```

Comment:
The default output for `codersdk.Workspace` would contain a `WorkspaceBuild` and `ProvisionerJob`, which could get *very lengthy*. We could not support "recursive" for tables but still print it for JSON and other formats.

Reply:
Any fields that don't have a table tag are never printed. We would just filter out what we don't care about by not adding tags for those fields

Reply:
Ahh, I think that's reasonable!